### PR TITLE
[RFC] Move HDFS support to separate repo

### PR DESCRIPTION
**Pull request by riversand963**
This PR moves HDFS support from RocksDB repo to a separate repo. The new (temporary?) repo
in this PR serves as an example before we finalize the decision on where and who to host hdfs support. At this point,
people can start from the example repo and fork.

Java/JNI is not included yet, and needs to be done later if necessary.

The goal is to include this commit in RocksDB 7.0 release.

Reference:
https://github.com/ajkr/dedupfs by @ajkr 

Test plan:
Follow the instructions in https://github.com/riversand963/rocksdb-hdfs-env/blob/master/README.md. Build and run db_bench and db_stress.

make check

### CODE REVIEW

**Comment by riversand963** [comment]
@adamretter, is it possible to move HDFS-related JNI code from main repo to https://github.com/riversand963/rocksdb-hdfs-env as well?

File: Makefile
Comment (mrambacher) [inline Makefile:253]: Shouldn't this be PLATFORM_CXXFLAGS?

**Review by mrambacher — CHANGES_REQUESTED** [review]
While I agree with most of this change, is there a reason we should move HDFS under plugins/hdfs in the main repo?  HDFS is one (of several) plugins that are currently in the source that should be moved into their own "untested" and "unsupported" place (like LUA and Rados and Cassandra).  

For those tools that currently use them, it would make it simpler if we kept the code in one place.  It would also simplify testing that we can do the right thing in building the plugins and getting the infrastructure working (regardless of if we ever actually RUN them).

**Review by ajkr — COMMENTED** [review]
The example repro should register a name matching the whole URI like "hdfs://.*" instead of "hdfs". Other than that it worked great!

We can clarify the temporary repo is meant as a starting point for whoever wants to take ownership assuming that is the intent.

**Comment by riversand963** [comment]
@mrambacher 
Just to clarify, we are not simply moving hdfs to plugins/hdfs in the main repo. We are removing hdfs support from main repo and moving it to a separate repo. Currently, it is hosted temporarily at https://github.com/riversand963/rocksdb-hdfs-env.git, and we can finalize the location or anybody can fork it. Future contribution to hdfs-specific code should not go to main repo. Only if somebody wants to build RocksDB with hdfs will they clone the repo to plugins/hdfs. We probably should do the same for all special Env/FS, e.g. rados, etc.

After removing the hdfs-specific code from main repo and ensuring there is a workaround solution as described in this PR, I currently do not have strong opinion on next step w.r.t #7949, and I don't think this PR should be blocked on a WIP feature. Even after HdfsEnv or HdfsFS and/or Env becomes Customizable, I think the approach in this PR should still work, and it does not prevent the development of #7949.

**Review by ajkr — COMMENTED** [review]
Looks like Java bindings need to be deleted too.

**Comment by adamretter** [comment]
@ajkr Yes indeed, it is on my TODO list (nearish the top)

**Comment by riversand963** [comment]
Hi @adamretter,
Discussed with @ajkr offline. Is it possible to expose a `createEnvFromUri()` public Java API? If so, is it possible to get rid of the Java binding code for HDFS?

**Comment by darionyaphet** [comment]
Maybe HDFS is not the only option. Would support be more optional in the future?

**Comment by adamretter** [comment]
> Is it possible to expose a createEnvFromUri() public Java API? 

@riversand963 As part of HDFS support, or as something else in RocksDB?

**Comment by mrambacher** [comment]
The Customizable infrastructure (e.g. XXX::CreateFromString) will work for all of the extension classes (Env is still pending).  I believe that is the interface that needs to be added to Java, for all of the Customizable classes

Having said that, I am not sure I understand how HDFS (or other plugins) can be accessed in Java unless they are built into the initial shared library, which means we will either be distributing a lot of Maven packages for the possible combinations (2 plugins means 4 packages, 3 means 8, etc) or have a means of dynamically adding them to the Java project.

**Comment by adamretter** [comment]
@mrambacher The approach taken in https://github.com/riversand963/rocksdb-hdfs-env/ is to build everything from source, so I doubt we will distribute those binaries. Otherwise with that approach we would have one complete binary of RocksDB for each plugin - and then what happens if you want two plugins?!?

**Comment by riversand963** [comment]
@adamretter correct, I don't think we will distribute binaries for different combinations of RocksDB + (HDFS + RADOS).

> As part of HDFS support, or as something else in RocksDB

Something else in RocksDB as a public Java API.

**Review by ajkr — APPROVED** [review]
LGTM, just need to remove Java binding.

**Comment by riversand963** [comment]
Thanks @ajkr for the review! Yeah, we will merge this after Java bindings for HDFS are deleted from main repo and we have an alternative.

**Comment by riversand963** [comment]
> Maybe HDFS is not the only option. Would support be more optional in the future?

@darionyaphet Not sure if I understand. Do you mean "support for more file systems"?

**Comment by darionyaphet** [comment]
> > Maybe HDFS is not the only option. Would support be more optional in the future?
> 
> @darionyaphet Not sure if I understand. Do you mean "support for more file systems"?

yep such as alluxio

**Comment by adamretter** [comment]
I am happy with what you are proposing. My concern is, if somebody wants two plugins, then your design is impossible... Or did I misunderstand?

**Comment by riversand963** [comment]
@darionyaphet no, we do not have such plans at the moment.

**Comment by riversand963** [comment]
@adamretter 
> I am happy with what you are proposing. My concern is, if somebody wants two plugins, then your design is impossible... Or did I misunderstand?

I assume you are replying to my prior comment.
I am not sure I follow. Can you elaborate on why it's not possible to have two plugins? Is this something specific in Java? In this PR, I can build RocksD with both dedupfs and hdfs and running db_stress with either of them.

**Comment by ajkr** [comment]
> @adamretter
> 
> > I am happy with what you are proposing. My concern is, if somebody wants two plugins, then your design is impossible... Or did I misunderstand?
> 
> I assume you are replying to my prior comment. I am not sure I follow. Can you elaborate on why it's not possible to have two plugins? Is this something specific in Java? In this PR, I can build RocksDB with both dedupfs and hdfs and run db_stress with either of them.

Right it is possible for users who build from source to include as many plugins as they want. IMO the Java release should include zero plugins. I suspect the Java release already does not include a functioning HdfsEnv since (before this PR) that required compiling with `USE_HDFS=1` and linking `-lhdfs`.

**Comment by riversand963** [comment]
@ajkr correct, current Java release does not include HDFS.

**Comment by riversand963** [comment]
Hi @alanpaxton, I have merged this PR to main, would it be possible for you to rebase https://github.com/riversand963/rocksdb/pull/4 onto facebook/main? Thanks!
cc @adamretter

**Pull request closed without merging**
